pkg/cloudprovider/providers/aws/aws_metrics.go

```diff
@@ -53,8 +71,94 @@ func recordAWSThrottlesMetric(operation string) {
 	awsAPIThrottlesMetric.With(prometheus.Labels{"operation_name": operation}).Inc()
 }
 
+func newVolumeStateCollector(cloud *Cloud) *volumeStateCollector {
+	return &volumeStateCollector{cloud}
+}
+
+type volumeStateCollector struct {
+	c *Cloud
+}
+
+var _ prometheus.Collector = &volumeStateCollector{}
+
+func (vsc *volumeStateCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- awsVolumesInStateTotal
+	ch <- awsVolumesInStateDurationMinutes
+}
+
+func (vsc *volumeStateCollector) Collect(ch chan<- prometheus.Metric) {
+	filters := []*ec2.Filter{newEc2Filter("instance-state-name", "running")}
+	instances, err := vsc.c.describeInstances(filters)
+	if err != nil {
+		glog.Warningf("Failed to describe instances while collecting volume state metrics: %v", err)
+		return
+	}
+
+	for volume, durations := range getVolumesInStateDuration(instances) {
+		for state, duration := range durations {
+			metric, err := prometheus.NewConstMetric(awsVolumesInStateDurationMinutes, prometheus.CounterValue, duration, volume, state)
+			if err != nil {
+				glog.Warningf("Failed to create metric for volume %q: %v", volume, err)
+			}
+			ch <- metric
+		}
+	}
+	for node, totals := range getVolumesInStateTotal(instances) {
+		for state, total := range totals {
+			metric, err := prometheus.NewConstMetric(awsVolumesInStateTotal, prometheus.GaugeValue, total, node, state)
+			if err != nil {
+				glog.Warningf("Failed to create metric for node %q: %v", node, err)
+			}
+			ch <- metric
+		}
+	}
+}
+
+func getVolumesInStateDuration(instances []*ec2.Instance) map[string]map[string]float64 {
+	durationsByVolume := make(map[string]map[string]float64)
+	for _, instance := range instances {
+		for _, mapping := range instance.BlockDeviceMappings {
+			ebs := mapping.Ebs
+			volume := string(awsVolumeID(aws.StringValue(ebs.VolumeId)))
+			state := *ebs.Status
+			attachTime := *ebs.AttachTime
+
+			switch state {
+			case "attaching", "detaching", "busy":
```

Comment:
Using value of `attachTime` for computing time since "busy" or "detaching" seems incorrect. Unfortunately this means, we gotta store absolute detach time somewhere..